File: exploration_main/src/main_exp.py
#! /usr/bin/env python

import os 
import time 
from math import fabs, pi
import actionlib
import roslaunch 
import rospy
from frontier_exploration.msg import ExploreTaskAction, ExploreTaskGoal
from geometry_msgs.msg import PoseStamped, Twist
from geometry_msgs.msg import PointStamped, Point, Point32, PolygonStamped
from geometry_msgs.msg import Vector3
from move_base_msgs.msg import MoveBaseAction, MoveBaseGoal
from nav_msgs.msg import Odometry
from std_msgs.msg import ColorRGBA, Header
from tf.transformations import euler_from_quaternion
from visualization_msgs.msg import Marker, MarkerArray
from exploration_perception.msg import DangerSign

PENDING = 0
ACTIVE = 1
DONE = 2
WARN = 3
ERROR = 4

class TurtlebotExploration:

    def __init__(self):
        #Initializes node, publishers and subscribers 
        rospy.init_node('map_navigation')
        self.movement_pub = rospy.Publisher('/cmd_vel_mux/input/teleop', Twist, queue_size=1)
        self.boundary_sub = rospy.Subscriber('/exploration_polygon_marker', Marker, self.boundary_callback)
        self.sign_sub = rospy.Subscriber('/danger_signs', DangerSign, self.sign_callback) 
        self.odom_sub = rospy.Subscriber('/odom', Odometry, self.odom_callback)
        self.marker_array_pub = rospy.Publisher("/visualization_marker_array", MarkerArray, queue_size=1)
        self.rate = rospy.Rate(20)

        #current robot pose coordinates updated by feedback callback function 
        self.pose = PoseStamped()
        self.x = 0.0
        self.y = 0.0
        self.yaw = 0.0

        #odometry yaw orientation
        self.odom_yaw = 0.0

        #current target frontier to explore 
        self.frontier = PoseStamped()   
        self.boundary_marker = Marker() 

        #danger sign message and boolean describing if detection happened
        self.danger_sign = DangerSign()
        self.sign_detected = False
        self.signs = MarkerArray()
    
    """
    This function initializes the explore_server action server and creates an action client object
    """
    def initialize_action_server(self, action_server_name, action):
        # Creates the SimpleActionClient
        client = actionlib.SimpleActionClient(action_server_name, action)
        # Waits until the action server has started up 
        rospy.loginfo('Waiting for action Server ' + action_server_name)
        client.wait_for_server()
        rospy.loginfo('Action Server Found...' + action_server_name)
        #return SimpleActionClient
        return client
  
    """
    This function will be called when feedback is received from the explore_server action server.
    It prints the current location of the robot based in the form of its x,y and yaw coordinates 
    and stores the location of the current frontier the robot is exploring.
    """
    def exploration_feedback_cb(self, feedback):
        self.pose = feedback.base_position
        self.x = feedback.base_position.pose.position.x
        self.y = feedback.base_position.pose.position.y
        c_or = feedback.base_position.pose.orientation
        (_, _, self.yaw) = euler_from_quaternion([c_or.x, c_or.y, c_or.z, c_or.w])
        self.frontier = feedback.next_frontier.pose
        rospy.loginfo("Feedback received")
        rospy.loginfo("Robot Position: x: {}, y: {}, yaw: {}".format(self.x, self.y, self.yaw))
        rospy.loginfo("Target Frontier: x: {}, y: {}".format(self.frontier.position.x, self.frontier.position.y))
    
    """
    This function will be called when feedback is received from the move base action server.
    It prints the current location of the robot based in the form of its x,y and yaw coordinates.
    """
    def movebase_feedback_cb(self, feedback):
        self.x = feedback.base_position.pose.position.x
        self.y = feedback.base_position.pose.position.y
        c_or = feedback.base_position.pose.orientation
        (_, _, self.yaw) = euler_from_quaternion([c_or.x, c_or.y, c_or.z, c_or.w])
        rospy.loginfo("Feedback received\nRobot Position: x: {}, y: {}, yaw: {}".format(self.x, self.y, self.yaw))
    
    """
    Subscriber to the '/danger_signs' topic; when a sign is detected it stores the DangerSign object and assigns the 
    boolean sign_detected a True value.
    """
    def sign_callback(self, data):
        self.danger_sign = data
        self.sign_detected = True
    
    """Subscriber to odom topic"""
    def odom_callback(self, data):
        orient = data.pose.pose.orientation
        (_, _, self.odom_yaw) = euler_from_quaternion([orient.x, orient.y, orient.z, orient.w])

    """Subscriber to exploration_boundary_polygon to show the exploration area defined"""
    def boundary_callback(self, data):
        self.boundary_marker = data

    """
    This function waits for at least for one subscriber to be active before publishing.
    It is useful when only one publish call is performed (usually) outside a while loop 
    """
    def publish_once(self, rotate):
        while not rospy.is_shutdown():
            connections = self.movement_pub.get_num_connections()
            if connections > 0:
                self.movement_pub.publish(rotate)
                break
            self.rate.sleep()
    
    """
    Rotate robot 360 degrees on the spot. 
    """
    def rotate360(self):
        move = Twist()
        target = 360
        target_in_radians = target * pi / 180

        #check when the 360 sweep has been finished
        while not rospy.is_shutdown():
            if self.odom_yaw < 0.0:
                current_yaw = self.odom_yaw + 2*pi
            else:
                current_yaw = self.odom_yaw
            speed = 0.5 * (target_in_radians - current_yaw)
            if speed < 0.01:
                break
            move.angular.z = speed
            self.movement_pub.publish(move)
            self.rate.sleep()
        move.angular.z = 0.0
        self.publish_once(move)

    """
    This function performs the turtlebot exploration of the unknown environment until no next frontier to explore is found.
    """
    def perform_exploration(self):
        #before the exploration starts, make robot perform an initial 360 sweep of the area 
        self.rotate360()
        
        #initialise action server and SimpleActionClient 
        exploration_server = '/explore_server'
        exploration_client = self.initialize_action_server(exploration_server, ExploreTaskAction)  

        #perform exploration
        goal = ExploreTaskGoal()

        #define boundaries of exploration 
        polygon_boundary = PolygonStamped()
        points = [Point32(x=-5.0, y=-5.0, z=0.0), Point32(x=5.0, y=-5.0, z=0.0), 
            Point32(x=5.0, y=5.0, z=0.0), Point32(x=-5.0, y=5.0, z=0.0), Point32(x=-5.0, y=-5.0, z=0.0)]
        polygon_boundary.polygon.points = points
        polygon_boundary.header.frame_id = 'map'
        polygon_boundary.header.stamp = rospy.Time.now()
        
        #centre point for frontier exploration, inside explore_boundary 
        centre_point = PointStamped()
        point = Point(x=0.0, y=0.0, z=0.0)
        centre_point.point = point
        centre_point.header.frame_id = 'map'
        centre_point.header.stamp = rospy.Time.now()

        goal.explore_boundary = polygon_boundary
        goal.explore_center = centre_point

        #Creates an rviz marker for the goal
        frontier_marker = Marker(type=Marker.CUBE, action=Marker.ADD, scale=Vector3(0.35, 0.35, 0.35),
            header=Header(frame_id='map'), color=ColorRGBA(1.0, 0.0, 0.0, 1.0))
        frontier_marker.pose = self.frontier

        #create marker array and append the markers to display
        marker_array = MarkerArray()
        marker_array.markers.append(self.boundary_marker, frontier_marker)

        #create the rviz marker for the signs 
        sign_marker = Marker(type=Marker.CUBE, action=Marker.ADD, scale=Vector3(0.75, 0.75, 0.05),
            header=Header(frame_id='map'), color=ColorRGBA(1.0, 0.0, 0.0, 1.0))

        # Sends the goal to the action server.
        rospy.loginfo('Sending goal to action server: %s', goal)
        exploration_client.send_goal(goal, feedback_cb=self.exploration_feedback_cb)

        # state_result will give the FINAL STATE. Will be 1 when Active, and 2 if NO ERROR, 3 If Any Warning, and 3 if ERROR
        state_result = exploration_client.get_state()

        rospy.loginfo("state_result: "+str(state_result))

        while state_result < DONE:
            rospy.loginfo("Checking for danger signs while performing exploration....")

            if self.sign_detected:
                sign_marker.pose = self.danger_sign.sign_pose.pose
                self.signs.markers.append(sign_marker)
                marker_array.markers.append(sign_marker)
                self.sign_detected = False
            
            #publish rviz markers for boundary polygon, frontier, signs 
            self.marker_array_pub.publish(marker_array)
                
            #update state value and display it on the log 
            state_result = exploration_client.get_state()
            rospy.loginfo("state_result: "+str(state_result))
            self.rate.sleep()

        rospy.loginfo("[Result] State: "+str(state_result))
        if state_result == ERROR:
            rospy.logerr("Something went wrong in the Server Side")
        if state_result == WARN:
            rospy.logwarn("There is a warning in the Server Side")
        else:
          # Waits for the server to finish performing the action.
          print 'Waiting for result...'
          exploration_client.wait_for_result()
        rospy.loginfo("Frontier Reached! Success!")
    
    """
    put description, locations are PoseStamped messages
    """
    def check_locations(self, locations):
         #initialise action server and SimpleActionClient 
        movebase_server = '/move_base'
        movebase_client = self.initialize_action_server(movebase_server, MoveBaseAction)  

        #create marker array and add all signs to it 
        marker_array = MarkerArray()
        marker_array.markers.extend(self.signs.markers)        

        for location in locations:
            # Creates a goal to send to the action server.
            goal = MoveBaseGoal()
            goal.target_pose.pose = location.pose
            goal.target_pose.header.frame_id = 'map'
            goal.target_pose.header.stamp = rospy.Time.now()

            #create goal marker 
            goal_marker = Marker(type=Marker.CUBE, action=Marker.ADD, scale=Vector3(0.35, 0.35, 0.35),
            header=Header(frame_id='map'), color=ColorRGBA(0.0, 1.0, 0.0, 1.0))
            goal_marker.pose = location.pose

            #add goal marker to marker array
            marker_array.markers.append(goal_marker)

            #create the rviz marker for the signs 
            sign_marker = Marker(type=Marker.CUBE, action=Marker.ADD, scale=Vector3(0.75, 0.75, 0.05),
                header=Header(frame_id='map'), color=ColorRGBA(1.0, 0.0, 0.0, 1.0))

            # Sends the goal to the action server.
            rospy.loginfo('Sending goal to action server: %s', goal)
            movebase_client.send_goal(goal, feedback_cb=self.movebase_feedback_cb)

            #perform navigation
            state_result = movebase_client.get_state()

            rospy.loginfo("state_result: "+str(state_result))
            if state_result == PENDING:
                rospy.loginfo("Action goal is pending...")
            if state_result == ACTIVE:
                rospy.loginfo("Action goal is being performed...")
                
            while state_result < DONE:
                rospy.loginfo("Checking for danger signs while performing navigation....")

                if self.sign_detected:
                    sign_marker.pose = self.danger_sign.sign_pose.pose
                    marker_array.markers.append(sign_marker)
                    self.sign_detected = False

                #publish rviz marker for goal
                self.marker_array_pub.publish(marker_array)

                #update state value and display it on the log 
                state_result = movebase_client.get_state()
                rospy.loginfo("state_result: "+str(state_result))
                self.rate.sleep()
            
            #wait for the server to finish performing the action
            movebase_client.wait_for_result()
            rospy.loginfo("Location Reached! Success!")

#create TurtlebotExploration object that initialises everything and start exploration
turtlebot_exploration = TurtlebotExploration()
turtlebot_exploration.perform_exploration()

#save the map into a file
uuid = roslaunch.rlutil.get_or_generate_uuid(None, False)
roslaunch.configure_logging(uuid)
#configure and execute the launch file 
launch = roslaunch.parent.ROSLaunchParent(uuid, ["/home/valeriofranchi/catkin_ws/src/exploration_main/launch/map_saver.launch"])
launch.start()

#define the map_saver generated file directories  
map_name = rospy.get_param("~map_name")
yaml_file_path = map_name + ".yaml"
pgm_file_path = map_name + ".pgm"
yaml_dir = os.path.join("home/valeriofranchi/catkin_ws/src/exploration_main", yaml_file_path)
pgm_dir = os.path.join("home/valeriofranchi/catkin_ws/src/exploration_main", pgm_file_path)

#if both files exist, wait for 3 seconds then stop the launch file 
if os.path.exists(yaml_dir) and os.path.exists(pgm_dir):
    time.sleep(3.0) 
    launch.shutdown()

File: exploration_main/src/mainexp.cpp
#include <geometry_msgs/PoseStamped.h>
#include <geometry_msgs/Twist.h>
#include <ros/ros.h>
#include <vector>
#include <tf/transform_datatypes.h>
#include <geometry_msgs/PolygonStamped.h>
#include <geometry_msgs/PointStamped.h>
#include <geometry_msgs/Point32.h>
#include <geometry_msgs/Point.h>
#include <frontier_exploration/ExploreTaskAction.h>
#include <frontier_exploration/ExploreTaskGoal.h>
#include <frontier_exploration/ExploreTaskFeedback.h>
#include <actionlib/client/simple_action_client.h>
#include <move_base_msgs/MoveBaseAction.h>
#include <move_base_msgs/MoveBaseGoal.h>
#include <move_base_msgs/MoveBaseFeedback.h>
#include <exploration_perception/DangerSign.h>
#include <geometry_msgs/Vector3.h>
#include <visualization_msgs/Marker.h>
#include <std_msgs/Header.h>
#include <std_msgs/ColorRGBA.h>
#include "exp.hpp"
#include "map_saver.cpp"

// function initialises the publishers for the TurtlebotExploration Class
void TurtlebotExploration::initialisePublishers() 
{
  velocityPublisher = nh.advertise<geometry_msgs::Twist>("/cmd_vel", 1);
  markerPublisher = nh.advertise<visualization_msgs::Marker>("/visualization_marker", 1);
}

// function initialises the subscribers for the TurtlebotExploration Class
void TurtlebotExploration::initialiseSubscribers() 
{
  signSubscriber = nh.subscribe<exploration_perception::DangerSign>("/danger_signs", 1, &TurtlebotExploration::signCB, this);
}

// definition of done callback. it is called once when the goal completes
void TurtlebotExploration::doneCB(const actionlib::SimpleClientGoalState& state) 
{
  ROS_INFO("[State Result]: %s", state.toString().c_str());
  ROS_INFO("The action has been completed");
  ros::shutdown();
}

// definition of the feedback callback. it will be called when the feedback is
// received from the action server. it just prints a message indicating a new
// message has been received

void TurtlebotExploration::moveBaseFeedbackCB(const move_base_msgs::MoveBaseFeedbackConstPtr& feedback) 
{
  x = feedback->base_position.pose.position.x;
  y = feedback->base_position.pose.position.y;
  tf::Quaternion q(feedback->base_position.pose.orientation.x, feedback->base_position.pose.orientation.y, 
    feedback->base_position.pose.orientation.z, feedback->base_position.pose.orientation.w);
  tf::Matrix3x3 m(q);
  double roll, pitch;
  m.getRPY(roll, pitch, yaw);
  ROS_INFO("[Feedback] Robot Position: x: %f, y: %f, yaw: %f", x, y, yaw);
}

// definition of the feedback callback. it will be clled when the feedback is
// received from the action server. it just prints a message indicating a new
// message has been received
void TurtlebotExploration::explorationFeedbackCB(const frontier_exploration::ExploreTaskFeedbackConstPtr& feedback) 
{
  x = feedback->base_position.pose.position.x;
  y = feedback->base_position.pose.position.y;
  tf::Quaternion q(feedback->base_position.pose.orientation.x, feedback->base_position.pose.orientation.y, 
    feedback->base_position.pose.orientation.z, feedback->base_position.pose.orientation.w);
  tf::Matrix3x3 m(q);
  double roll, pitch;
  m.getRPY(roll, pitch, yaw);
  ROS_INFO("[Feedback] Robot Position: x: %f, y: %f, yaw: %f", x, y, yaw);
  frontier = feedback->next_frontier.pose;
  ROS_INFO("[Feedback] Target Frontier: x: %f, y: %f", frontier.position.x, frontier.position.y);
}

void TurtlebotExploration::signCB(const exploration_perception::DangerSignConstPtr& msg)
{
  dangerSign = *msg;
  signDetected = true;
}

void TurtlebotExploration::publishOnce(const geometry_msgs::Twist& vel) 
{
  ros::Rate rate(20);
  while (ros::ok()) {
    int numConnections = velocityPublisher.getNumSubscribers();
    if (numConnections > 0) {
      velocityPublisher.publish(vel);
      break;
    } else {
      rate.sleep();
    }
  }
}

void TurtlebotExploration::rotate360()
{
  geometry_msgs::Twist move;
  double init_yaw = yaw;
  move.angular.z = 0.1;

  //check when 360 sweep has finished
  while (1)
  {
    velocityPublisher.publish(move);
    if ((init_yaw > 0.0 && yaw < 0.0) || (init_yaw < 0.0 && yaw < init_yaw))
      yaw += 2 * M_PI;
    double difference = std::abs(init_yaw - yaw);
    if (difference > 2 * M_PI) 
      break;
  }

  //stop robot
  move.angular.z = 0.0;
  velocityPublisher.publish(move);
}

void TurtlebotExploration::performExploration() 
{
  //before exploration starts, make robot perform an initial 360 sweep of the area
  TurtlebotExploration::rotate360();

  //initialize action server 
  typedef actionlib::SimpleActionClient<frontier_exploration::ExploreTaskAction> ExplorationClient;
  ExplorationClient explorationClient("/explore_server", true);

  //wait for the action server to come up
  while(!explorationClient.waitForServer())
  {
    ROS_INFO("Waiting for the explore_server action server to come up");            
  }      
    ROS_INFO("explore_server action server is up");

  //create goal to send to the action server
  frontier_exploration::ExploreTaskGoal goal;

  //define boundaries of exploration
  geometry_msgs::PolygonStamped polygonBoundary;
  geometry_msgs::Point32 point1, point2, point3, point4;
  point1.x, point1.y, point1.z = -5.0, -5.0, 0.0;
  point2.x, point2.y, point2.z = 5.0, -5.0, 0.0;
  point3.x, point3.y, point3.z = 5.0, 5.0, 0.0;
  point4.x, point4.y, point4.z = -5.0, 5.0, 0.0;

  std::vector<geometry_msgs::Point32> points;
  points.reserve(5);
  points.push_back(point1);
  points.push_back(point2);
  points.push_back(point3);
  points.push_back(point4);
  points.push_back(point1);

  polygonBoundary.polygon.points = points;
  polygonBoundary.header.frame_id = "map";
  polygonBoundary.header.stamp = ros::Time::now();
  
  //centre point for exploration
  geometry_msgs::PointStamped centrePoint;
  geometry_msgs::Point point;
  point.x, point.y, point.z = 0.0, 0.0, 0.0;
  centrePoint.point = point;
  centrePoint.header.frame_id = "map";
  centrePoint.header.stamp = ros::Time::now();

  goal.explore_boundary = polygonBoundary;
  goal.explore_center = centrePoint;

  //create goal marker
  visualization_msgs::Marker frontierMarker;
  frontierMarker.type = visualization_msgs::Marker::CUBE;
  frontierMarker.action = visualization_msgs::Marker::ADD;
  geometry_msgs::Vector3 fscale;
  fscale.x, fscale.y, fscale.z = 0.35, 0.35, 0.35;
  frontierMarker.scale = fscale;
  std_msgs::Header fHeader;
  fHeader.frame_id = "map";
  frontierMarker.header = fHeader;
  std_msgs::ColorRGBA fcolour;
  fcolour.r, fcolour.g, fcolour.b, fcolour.a = 1.0, 0.0, 0.0, 1.0;
  frontierMarker.color = fcolour;
  frontierMarker.pose = frontier;

  //create sign marker
  visualization_msgs::Marker signMarker;
  signMarker.type = visualization_msgs::Marker::CUBE;
  signMarker.action = visualization_msgs::Marker::ADD;
  geometry_msgs::Vector3 sScale;
  sScale.x, sScale.y, sScale.z = 0.75, 0.75, 0.05;
  signMarker.scale = sScale;
  std_msgs::Header sHeader;
  sHeader.frame_id = "map";
  signMarker.header = sHeader;
  std_msgs::ColorRGBA sColour;
  sColour.r, sColour.g, sColour.b, sColour.a = 1.0, 0.0, 0.0, 1.0;
  signMarker.color = sColour;
  
  //send goal to action server 
  ROS_INFO("Sending goal to action server...");
  explorationClient.sendGoal(goal, boost::bind(&TurtlebotExploration::doneCB, this, _1), ExplorationClient::SimpleActiveCallback(),
    boost::bind(&TurtlebotExploration::explorationFeedbackCB, this, _1));
  
  //perform navigation
  ros::Rate rate(20);
  actionlib::SimpleClientGoalState stateResult = explorationClient.getState();
  ROS_INFO("[State Result]: %s", stateResult.toString().c_str());

  while ( stateResult == actionlib::SimpleClientGoalState::ACTIVE || 
    stateResult == actionlib::SimpleClientGoalState::PENDING )
  {
    ROS_INFO("Checking for danger signs while performing navigation...");

    //publish marker
    markerPublisher.publish(frontierMarker);

    if (signDetected)
    {
      signMarker.pose = dangerSign.sign_pose.pose;
      signDetected = false;
    }
    
    //update state value and display it on log
    stateResult = explorationClient.getState();
    ROS_INFO("[State Result]: %s", stateResult.toString().c_str());
    rate.sleep();
  }

  //wait for the server to finish performing the action
  explorationClient.waitForResult();
  ROS_INFO("Unknown Environment Explored! Success!");
}

void TurtlebotExploration::checkLocations(const std::vector<geometry_msgs::PoseStamped>& locations)
{
  //initialize action server 
  typedef actionlib::SimpleActionClient<move_base_msgs::MoveBaseAction> MoveBaseClient;
  MoveBaseClient mbClient("move_base", true);

  //wait for the action server to come up
  while(!mbClient.waitForServer())
  {
    ROS_INFO("Waiting for the move_base action server to come up");            
  }      
    ROS_INFO("move_base action server is up");

  std::vector<geometry_msgs::PoseStamped>::const_iterator constIter;
  for(constIter = locations.begin(); constIter != locations.end(); ++constIter)
  { 
    //create goal to send to the action server
    move_base_msgs::MoveBaseGoal goal;
    goal.target_pose.pose = (*constIter).pose;
    goal.target_pose.header.frame_id = "map";
    goal.target_pose.header.stamp = ros::Time::now();

    //create goal marker
    visualization_msgs::Marker goalMarker;
    goalMarker.type = visualization_msgs::Marker::CUBE;
    goalMarker.action = visualization_msgs::Marker::ADD;
    geometry_msgs::Vector3 gScale;
    gScale.x, gScale.y, gScale.z = 0.35, 0.35, 0.35;
    goalMarker.scale = gScale;
    std_msgs::Header gHeader;
    gHeader.frame_id = "map";
    goalMarker.header = gHeader;
    std_msgs::ColorRGBA gColour;
    gColour.r, gColour.g, gColour.b, gColour.a = 0.0, 1.0, 0.0, 1.0;
    goalMarker.color = gColour;
    goalMarker.pose = (*constIter).pose;

    //create sign marker
    visualization_msgs::Marker signMarker;
    signMarker.type = visualization_msgs::Marker::CUBE;
    signMarker.action = visualization_msgs::Marker::ADD;
    geometry_msgs::Vector3 sScale;
    sScale.x, sScale.y, sScale.z = 0.75, 0.75, 0.05;
    signMarker.scale = sScale;
    std_msgs::Header sHeader;
    sHeader.frame_id = "map";
    signMarker.header = sHeader;
    std_msgs::ColorRGBA sColour;
    sColour.r, sColour.g, sColour.b, sColour.a = 1.0, 0.0, 0.0, 1.0;
    signMarker.color = sColour;

    //send goal to action server 
    ROS_INFO("Sending goal to action server...");
    mbClient.sendGoal(goal, boost::bind(&TurtlebotExploration::doneCB, this, _1), MoveBaseClient::SimpleActiveCallback(),
      boost::bind(&TurtlebotExploration::moveBaseFeedbackCB, this, _1));

    //perform navigation
    ros::Rate rate(20);
    actionlib::SimpleClientGoalState stateResult = mbClient.getState();
    ROS_INFO("[State Result]: %s", stateResult.toString().c_str());
      
    while ( stateResult == actionlib::SimpleClientGoalState::ACTIVE || 
      stateResult == actionlib::SimpleClientGoalState::PENDING )
    {
      ROS_INFO("Checking for danger signs while performing navigation...");

      //publish marker
      markerPublisher.publish(goalMarker);

      if (signDetected)
      {
        signMarker.pose = dangerSign.sign_pose.pose;
        signDetected = false;
      }
      
      //update state value and display it on log
      stateResult = mbClient.getState();
      ROS_INFO("[State Result]: %s", stateResult.toString().c_str());
      rate.sleep();
    }

    //wait for the server to finish performing the action
    mbClient.waitForResult();
    ROS_INFO("Location Reached! Success!");
  }
}

int main(int argc, char** argv){
  ros::init(argc, argv, "turtlebot_exploration_node");
  ros::NodeHandle nodeHandle;

  //create TurtlebotExploration object and perform exploration 
  TurtlebotExploration turtlebotExp(&nodeHandle);
  //turtlebotExp.performExploration();

  //save map
  std::string mapname = "map";
  int threshold_occupied = 65;
  int threshold_free = 25;

  nodeHandle.getParam("~map_name", mapname);

  MapGenerator mg(mapname, threshold_occupied, threshold_free);

  while(!mg.saved_map_ && ros::ok())
    ros::spinOnce();
  
  return 0;
}